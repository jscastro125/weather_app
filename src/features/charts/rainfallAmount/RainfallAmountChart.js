import React, {useEffect} from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer} from 'recharts';

import {
    fetchRainfallAmounts,
    selectRainfallAmounts
} from './rainfallSlice';
import {getDayName} from "../../../utils/utils";

export default function RainfallAmount() {
    const rainfallAmounts = useSelector(selectRainfallAmounts);
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(fetchRainfallAmounts());
    }, [dispatch]);
    const data = rainfallAmounts.map(({ day, amount }) => ({ day: getDayName(day), amount }));
    return (
        <>
            <h3>
                Amount of rainfall
            </h3>
            <ResponsiveContainer width="100%" height={250}>
                <BarChart data={data}>
                    <CartesianGrid strokeDasharray="3 3" />
                    <XAxis dataKey="day" />
                    <YAxis />
                    <Tooltip />
                    <Bar dataKey="amount" fill="#000" />
                </BarChart>
            </ResponsiveContainer>
        </>
    );
}
