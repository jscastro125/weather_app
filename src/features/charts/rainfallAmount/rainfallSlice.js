import { createSlice } from '@reduxjs/toolkit';

export const rainfallSlice = createSlice({
    name: 'rainfallAmounts',
    initialState: {
        value: [],
        isLoading: false,
        error: null
    },
    reducers: {
        setIsLoading: (state, action) => {
            state.isLoading = action.payload;
        },
        setError: (state, action) => {
            state.error = action.payload;
        },
        setValue: (state, action) => {
            state.value = action.payload;
        },
    },
});

export const { setIsLoading, setError, setValue } = rainfallSlice.actions;

export const fetchRainfallAmounts = () => dispatch => {
    dispatch(setIsLoading(true));
    fetch('https://private-4945e-weather34.apiary-proxy.com/weather34/rain')
        .then(res => res.json())
        .then(data => {
            if(data?.length !== 0) {
                const [currentWeek] = data;
                dispatch(setValue(currentWeek?.days));
                dispatch(setIsLoading(false));
            }
        })
        .catch(err => dispatch(setError(err)))
};

export const selectRainfallAmounts = state => state.rainfallAmounts.value;
export const selectError = state => state.rainfallAmounts.selectError;

export default rainfallSlice.reducer;
