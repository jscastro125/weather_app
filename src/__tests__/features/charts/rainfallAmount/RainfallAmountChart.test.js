import React from "react";
import {useDispatch, useSelector} from 'react-redux';
import { create } from "react-test-renderer";

import * as amountsSlice from '../../../../features/charts/rainfallAmount/rainfallSlice';
import RainfallAmount from '../../../../features/charts/rainfallAmount/RainfallAmountChart';
import rainfallAmountsMock from '../../../../__mocks__/rainfallAmounts.mock';

jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
    useDispatch: jest.fn()
}));

function createNodeMock() {
    const doc = document.implementation.createHTMLDocument();
    return { parentElement: doc.body };
}

describe("<RainfallAmount />", () => {
    it("should match snapshot", () => {
        useSelector.mockImplementation(sel => sel());
        useDispatch.mockImplementation(() => () => {});
        const spyAmounts = jest.spyOn(amountsSlice, 'selectRainfallAmounts');
        spyAmounts.mockReturnValue(rainfallAmountsMock);
        window.innerWidth = 1100;
        const component = create(<RainfallAmount />, { createNodeMock });
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
});
