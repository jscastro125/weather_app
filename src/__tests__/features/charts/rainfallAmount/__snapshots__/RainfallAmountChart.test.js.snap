// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RainfallAmount /> should match snapshot 1`] = `
Array [
  <h3>
    Amount of rainfall
  </h3>,
  <div
    className="recharts-responsive-container"
    style={
      Object {
        "height": 250,
        "maxHeight": undefined,
        "minHeight": undefined,
        "minWidth": undefined,
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "display": "none",
          "height": 0,
          "position": "absolute",
          "visibility": "hidden",
          "width": 0,
        }
      }
    />
  </div>,
]
`;
